File: poc/kameleo-container-probe/probe.py
import json
import os
import sys
from urllib.parse import urlsplit

from kameleo.local_api_client import KameleoLocalApiClient
from kameleo.local_api_client.models import BrowserSettings, CreateProfileRequest
from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
from playwright.sync_api import sync_playwright


LOGIN_PATH = "/memapi/jaxrs/xt_login/agree/v1"


def redacted_url(value: str) -> str:
    parsed = urlsplit(value)
    return f"{parsed.scheme}://{parsed.netloc}{parsed.path}"


endpoint = os.environ.get("KAMELEO_ENDPOINT", "http://127.0.0.1:5050")
authenticate = "--auth" in sys.argv[1:]
keep_profile = os.environ.get("KEEP_PROFILE") == "1"
warmup = os.environ.get("WARMUP") == "1"
client = KameleoLocalApiClient(endpoint=endpoint)
client.verify_engine_ready()

profile_name = "kogane-vpass-persistent-windows-chrome"
profile = next(
    (item for item in client.profile.list_profiles() if item.name == profile_name),
    None,
)
if profile is None:
    fingerprints = client.fingerprint.search_fingerprints(
        device_type="desktop",
        os_family="windows",
        browser_product="chrome",
        browser_version=">145",
    )
    if not fingerprints:
        raise RuntimeError("no recent Windows Chrome fingerprint was returned")
    profile = client.profile.create_profile(
        CreateProfileRequest(
            fingerprintId=fingerprints[0].id,
            name=profile_name,
            language="ja-JP,ja",
        )
    )

result = {
    "engine": "kameleo-chroma",
    "requested_fingerprint": "windows/chrome/>145",
    "authenticate": authenticate,
    "persistent": keep_profile,
    "warmup": warmup,
}

try:
    # Cloudflare Containers does not expose Docker's --shm-size runtime knob.
    # Make Chromium use /tmp instead of the default 64 MiB /dev/shm mount.
    client.profile.start_profile(
        profile.id,
        BrowserSettings(arguments=["headless", "disable-dev-shm-usage"]),
        _request_timeout=180,
    )

    with sync_playwright() as playwright:
        browser = playwright.chromium.connect_over_cdp(
            f"ws://127.0.0.1:5050/playwright/{profile.id}", timeout=90_000
        )
        context = browser.contexts[0]
        page = context.pages[0] if context.pages else context.new_page()

        if warmup:
            page.goto(
                "https://www.smbc-card.com/",
                wait_until="domcontentloaded",
                timeout=45_000,
            )
            page.wait_for_timeout(2_000)
            page.mouse.move(180, 160, steps=12)
            page.mouse.wheel(0, 500)
            page.wait_for_timeout(1_200)
            page.mouse.move(740, 420, steps=18)
            page.mouse.wheel(0, -250)
            page.wait_for_timeout(1_500)

        trace_response = page.goto(
            "https://www.cloudflare.com/cdn-cgi/trace",
            wait_until="domcontentloaded",
            timeout=30_000,
        )
        trace_fields = dict(
            line.split("=", 1)
            for line in page.locator("body").inner_text().splitlines()
            if "=" in line
        )
        result["egress"] = {
            "status": trace_response.status if trace_response else None,
            "country": trace_fields.get("loc"),
            "colo": trace_fields.get("colo"),
            "warp": trace_fields.get("warp"),
        }

        result["runtime"] = page.evaluate(
            """() => {
              const canvas = document.createElement('canvas');
              const gl = canvas.getContext('webgl');
              const debug = gl?.getExtension('WEBGL_debug_renderer_info');
              return {
                userAgent: navigator.userAgent,
                platform: navigator.platform,
                languages: navigator.languages,
                webdriver: navigator.webdriver,
                screen: {
                  width: screen.width,
                  height: screen.height,
                  colorDepth: screen.colorDepth,
                },
                hardwareConcurrency: navigator.hardwareConcurrency,
                deviceMemory: navigator.deviceMemory ?? null,
                maxTouchPoints: navigator.maxTouchPoints,
                webglVendor: debug ? gl.getParameter(debug.UNMASKED_VENDOR_WEBGL) : null,
                webglRenderer: debug ? gl.getParameter(debug.UNMASKED_RENDERER_WEBGL) : null,
              };
            }"""
        )

        login_page_response = page.goto(
            "https://www.smbc-card.com/mem/index.jsp",
            wait_until="domcontentloaded",
            timeout=45_000,
        )
        body = page.locator("body").inner_text(timeout=10_000)
        result["vpass"] = {
            "status": login_page_response.status if login_page_response else None,
            "url": redacted_url(page.url),
            "title": page.title(),
            "login_form": page.locator("#id_input").count() > 0
            and page.locator("#pw_input").count() > 0,
            "access_denied": "access denied"
            in f"{page.title()} {body[:2000]}".lower(),
        }

        if authenticate:
            print("READY_FOR_CREDENTIALS", file=sys.stderr, flush=True)
            user_id = sys.stdin.readline().rstrip("\r\n")
            password = sys.stdin.readline().rstrip("\r\n")
            if not user_id or not password:
                raise ValueError("two credential lines are required")

            id_input = page.locator("#id_input")
            password_input = page.locator("#pw_input")
            id_box = id_input.bounding_box()
            if id_box:
                page.mouse.move(
                    id_box["x"] + id_box["width"] / 2,
                    id_box["y"] + id_box["height"] / 2,
                    steps=24,
                )
                page.mouse.click(
                    id_box["x"] + id_box["width"] / 2,
                    id_box["y"] + id_box["height"] / 2,
                )
            else:
                id_input.click()
            page.keyboard.type(user_id, delay=115)
            page.wait_for_timeout(450)

            password_box = password_input.bounding_box()
            if password_box:
                page.mouse.move(
                    password_box["x"] + password_box["width"] / 2,
                    password_box["y"] + password_box["height"] / 2,
                    steps=18,
                )
                page.mouse.click(
                    password_box["x"] + password_box["width"] / 2,
                    password_box["y"] + password_box["height"] / 2,
                )
            else:
                password_input.click()
            page.keyboard.type(password, delay=95)
            result["interaction"] = {
                "id_length": len(id_input.input_value()),
                "password_length": len(password_input.input_value()),
            }
            user_id = ""
            password = ""
            page.wait_for_timeout(700)

            login_button = page.get_by_role("button", name="ログイン", exact=True)
            login_button.scroll_into_view_if_needed()
            page.wait_for_timeout(250)
            login_box = login_button.bounding_box()
            if login_box:
                page.mouse.move(
                    login_box["x"] + login_box["width"] / 2,
                    login_box["y"] + login_box["height"] / 2,
                    steps=28,
                )
                page.wait_for_timeout(350)

            login_response = None
            try:
                with page.expect_response(
                    lambda response: urlsplit(response.url).path == LOGIN_PATH,
                    timeout=30_000,
                ) as response_info:
                    if login_box:
                        page.mouse.click(
                            login_box["x"] + login_box["width"] / 2,
                            login_box["y"] + login_box["height"] / 2,
                        )
                    else:
                        login_button.click()
                login_response = response_info.value
            except PlaywrightTimeoutError:
                pass

            page.wait_for_timeout(5_000)
            body = page.locator("body").inner_text(timeout=10_000)
            result["login"] = {
                "response_status": login_response.status if login_response else None,
                "response_url": redacted_url(login_response.url)
                if login_response
                else None,
                "login_result_header": login_response.headers.get("x-loginresult")
                if login_response
                else None,
                "final_url": redacted_url(page.url),
                "title": page.title(),
                "authenticated": page.get_by_role(
                    "link", name="ログアウト", exact=True
                ).count()
                > 0
                or "操作中のカードを変更する" in body,
                "blocked": (
                    login_response is not None and login_response.status == 403
                )
                or "access denied" in body.lower(),
            }

        browser.close()
finally:
    try:
        client.profile.stop_profile(profile.id, _request_timeout=60)
    except Exception:
        pass
    if not keep_profile:
        try:
            client.profile.delete_profile(profile.id, _request_timeout=60)
        except Exception:
            pass

print(json.dumps(result, ensure_ascii=False, indent=2))

File: poc/oci-browser-probe/probe.mjs
import { createInterface } from "node:readline";
import { chromium } from "playwright";

const modes = new Set(["chrome-launch", "chrome-cdp", "chromium-launch"]);
const mode = process.argv[2];
const authenticate = process.argv.includes("--auth");

if (!modes.has(mode)) {
  console.error("usage: node probe.mjs <chrome-launch|chrome-cdp|chromium-launch> [--auth]");
  process.exit(2);
}

const redactUrl = (value) => {
  try {
    const url = new URL(value);
    return `${url.origin}${url.pathname}`;
  } catch {
    return "invalid-url";
  }
};

const readCredentials = async () => {
  console.error("READY_FOR_CREDENTIALS");
  const input = createInterface({ input: process.stdin, crlfDelay: Infinity, terminal: false });
  const iterator = input[Symbol.asyncIterator]();
  const id = (await iterator.next()).value;
  const password = (await iterator.next()).value;
  input.close();
  if (!id || !password) throw new Error("two credential lines are required");
  return { id, password };
};

const launch = async () => {
  if (mode === "chrome-cdp") {
    const endpoint = process.env.KOGANE_CDP_URL ?? "http://127.0.0.1:9222";
    const browser = await chromium.connectOverCDP(endpoint);
    const context = browser.contexts()[0];
    if (!context) throw new Error("CDP browser has no default context");
    return { browser, context };
  }

  const options = { headless: true, args: ["--disable-dev-shm-usage"] };
  if (mode === "chrome-launch") options.executablePath = "/usr/bin/google-chrome-stable";
  const browser = await chromium.launch(options);
  const context = await browser.newContext({ locale: "ja-JP", timezoneId: "Asia/Tokyo" });
  return { browser, context };
};

const result = { timestamp: new Date().toISOString(), mode, authenticate };

let browser;
let context;
let page;
let credentials;

try {
  ({ browser, context } = await launch());
  result.browserVersion = browser.version();
  page = await context.newPage();

  const traceResponse = await page.goto("https://www.cloudflare.com/cdn-cgi/trace", {
    waitUntil: "domcontentloaded",
    timeout: 30_000,
  });
  if (!traceResponse?.ok()) throw new Error(`Cloudflare trace failed: ${traceResponse?.status()}`);
  const trace = await page.locator("body").innerText();
  const fields = Object.fromEntries(
    trace
      .split("\n")
      .map((line) => line.split("=", 2))
      .filter((pair) => pair.length === 2),
  );
  result.egress = { ip: fields.ip, loc: fields.loc, warp: fields.warp, gateway: fields.gateway };

  result.runtime = await page.evaluate(() => ({
    userAgent: navigator.userAgent,
    platform: navigator.platform,
    webdriver: navigator.webdriver,
    languages: navigator.languages,
  }));

  const loginPageResponse = await page.goto("https://www.smbc-card.com/mem/index.jsp", {
    waitUntil: "domcontentloaded",
    timeout: 45_000,
  });
  result.bootstrap = {
    status: loginPageResponse?.status(),
    url: redactUrl(page.url()),
    title: await page.title(),
    hasIdInput:
      (await page
        .locator("input#id_input, input[placeholder*='ID'], input[aria-label*='ID']")
        .count()) > 0,
    hasPasswordInput: (await page.locator("input[type='password']").count()) > 0,
  };

  if (authenticate) {
    credentials = await readCredentials();
    const idInput = page
      .locator("input#id_input, input[placeholder*='ID'], input[aria-label*='ID']")
      .first();
    const passwordInput = page.locator("input#pw_input, input[type='password']").first();
    await idInput.fill(credentials.id);
    await passwordInput.fill(credentials.password);
    credentials.id = "";
    credentials.password = "";
    credentials = undefined;

    const loginResponsePromise = page
      .waitForResponse(
        (response) => {
          try {
            return new URL(response.url()).pathname === "/memapi/jaxrs/xt_login/agree/v1";
          } catch {
            return false;
          }
        },
        { timeout: 30_000 },
      )
      .catch(() => undefined);
    await page.getByRole("button", { name: "ログイン", exact: true }).click();
    const loginResponse = await loginResponsePromise;
    await page.waitForTimeout(5_000);

    const bodyText = await page
      .locator("body")
      .innerText()
      .catch(() => "");
    result.login = {
      responseStatus: loginResponse?.status(),
      responseUrl: loginResponse ? redactUrl(loginResponse.url()) : undefined,
      loginResultHeader: loginResponse?.headers()["x-loginresult"],
      finalUrl: redactUrl(page.url()),
      title: await page.title(),
      authenticated:
        (await page.getByRole("link", { name: "ログアウト", exact: true }).count()) > 0 ||
        bodyText.includes("操作中のカードを変更する"),
      blocked: loginResponse?.status() === 403 || bodyText.includes("Access Denied"),
    };
  }

  console.log(JSON.stringify(result, null, 2));
} catch (error) {
  result.error = error instanceof Error ? error.message : String(error);
  console.log(JSON.stringify(result, null, 2));
  process.exitCode = 1;
} finally {
  if (credentials) {
    credentials.id = "";
    credentials.password = "";
  }
  await page?.close().catch(() => {});
  await browser?.close().catch(() => {});
}
